File: ios/BitchatModule.swift
import Foundation

@objc(Bitchat)
class BitchatModule: NSObject {
    private let channel = ChannelManager()
    private let bleService = BleService()
    private let encryption = Encryption()
    private var messageCallback: RCTResponseSenderBlock?
    private var peerConnectedCallback: RCTResponseSenderBlock?
    private var peerDisconnectedCallback: RCTResponseSenderBlock?
    private var deliveryAckCallback: RCTResponseSenderBlock?
    private var readReceiptCallback: RCTResponseSenderBlock?
    private var deliveryStatusUpdateCallback: RCTResponseSenderBlock?
    private let favoritesKey = "BitchatFavorites"

    @objc func startAdvertising(peerID: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        guard peerID.count == 8 else {
            rejecter("INVALID_PEER_ID", nil, NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Peer ID must be 8 bytes"]))
            return
        }
        do {
            try bleService.startAdvertising(peerID: peerID)
            try bleService.sendCoverTraffic(peerID: peerID) // Include cover traffic
            resolver(nil)
        } catch {
            rejecter("ADVERTISE_ERROR", nil, error)
        }
    }

    @objc func scanPeers(resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        do {
            let peers = try bleService.scanPeers()
            resolver(peers.map { $0 as Any })
        } catch {
            rejecter("SCAN_ERROR", nil, error)
        }
    }

    @objc func sendMessage(message: [String: Any], recipient: String?, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        guard let msg = mapToBitchatMessage(message) else {
            rejecter("INVALID_MESSAGE", nil, NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid message format"]))
            return
        }
        do {
            msg.deliveryStatus = "PENDING"
            if let recipient = recipient { msg.recipientNickname = recipient }
            try bleService.sendMessage(msg, recipient: recipient)
            resolver(nil)
        } catch {
            rejecter("SEND_ERROR", nil, error)
        }
    }

    @objc func sendEncryptedChannelMessage(message: [String: Any], channel: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        guard let msg = mapToBitchatMessage(message) else {
            rejecter("INVALID_MESSAGE", nil, NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid message format"]))
            return
        }
        do {
            let key = encryption.deriveChannelKey(channel, channel: channel) // Use channel as password
            if let content = msg.content {
                msg.encryptedContent = encryption.encryptContent(content, key: key)
                msg.isEncrypted = true
                msg.channel = channel
                try bleService.sendEncryptedChannelMessage(msg, channel: channel)
                resolver(nil)
            } else {
                rejecter("NO_CONTENT", nil, NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Message content required"]))
            }
        } catch {
            rejecter("ENCRYPT_SEND_ERROR", nil, error)
        }
    }

    @objc func onMessageReceived(callback: @escaping RCTResponseSenderBlock) {
        messageCallback = callback
    }

    @objc func onPeerConnected(callback: @escaping RCTResponseSenderBlock) {
        peerConnectedCallback = callback
    }

    @objc func onPeerDisconnected(callback: @escaping RCTResponseSenderBlock) {
        peerDisconnectedCallback = callback
    }

    @objc func onDeliveryAck(callback: @escaping RCTResponseSenderBlock) {
        deliveryAckCallback = callback
    }

    @objc func onReadReceipt(callback: @escaping RCTResponseSenderBlock) {
        readReceiptCallback = callback
    }

    @objc func onDeliveryStatusUpdate(callback: @escaping RCTResponseSenderBlock) {
        deliveryStatusUpdateCallback = callback
    }

    @objc func decryptChannelMessage(encryptedContent: String, channel: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        do {
            let key = encryption.deriveChannelKey(channel, channel: channel)
            let decrypted = try encryption.decryptContent(Data(encryptedContent.utf8), key: key)
            resolver(String(data: decrypted, encoding: .utf8))
        } catch {
            rejecter("DECRYPT_ERROR", nil, error)
        }
    }

    @objc func isFavorite(fingerprint: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        guard !fingerprint.isEmpty else {
            rejecter("INVALID_FINGERPRINT", nil, NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Fingerprint cannot be empty"]))
            return
        }
        let favorites = UserDefaults.standard.array(forKey: favoritesKey) as? [String] ?? []
        let isFavorite = favorites.contains(fingerprint)
        resolver(isFavorite)
    }

    @objc func joinChannel(channel: String, password: String?, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        do {
            let success = channel.joinChannel(channel, password: password, peerID: "peerID")
            resolver(success)
        } catch {
            rejecter("JOIN_ERROR", nil, error)
        }
    }

    @objc func setChannelPassword(channel: String, password: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        do {
            channel.setChannelPassword(channel, password: password, peerID: "peerID")
            resolver(nil)
        } catch {
            rejecter("SET_PASSWORD_ERROR", nil, error)
        }
    }

    @objc func removeChannelPassword(channel: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        do {
            channel.removeChannelPassword(channel, peerID: "peerID")
            resolver(nil)
        } catch {
            rejecter("REMOVE_PASSWORD_ERROR", nil, error)
        }
    }

    @objc func transferChannelOwnership(channel: String, newOwnerID: String, resolver: RCTPromiseResolveBlock, rejecter: RCTPromiseRejectBlock) {
        do {
            channel.transferOwnership(channel, newOwnerID: newOwnerID, peerID: "peerID")
            resolver(nil)
        } catch {
            rejecter("TRANSFER_ERROR", nil, error)
        }
    }

    private func mapToBitchatMessage(_ map: [String: Any]) -> BitchatMessage? {
        var msg = BitchatMessage()
        msg.id = map["id"] as? String
        msg.sender = map["sender"] as? String
        msg.content = map["content"] as? String
        msg.timestamp = (map["timestamp"] as? NSNumber)?.doubleValue ?? Date().timeIntervalSince1970 * 1000
        msg.isRelay = map["isRelay"] as? Bool ?? false
        msg.originalSender = map["originalSender"] as? String
        msg.isPrivate = map["isPrivate"] as? Bool ?? false
        msg.recipientNickname = map["recipientNickname"] as? String
        msg.senderPeerID = map["senderPeerID"] as? String
        msg.mentions = map["mentions"] as? [String] ?? []
        msg.channel = map["channel"] as? String
        msg.encryptedContent = map["encryptedContent"] as? Data
        msg.isEncrypted = map["isEncrypted"] as? Bool ?? false
        msg.deliveryStatus = map["deliveryStatus"] as? String ?? "PENDING"
        return msg
    }

    // Helper method to add a favorite (not exposed to JS, for internal use if needed)
    private func addFavorite(_ fingerprint: String) {
        var favorites = UserDefaults.standard.array(forKey: favoritesKey) as? [String] ?? []
        if !favorites.contains(fingerprint) {
            favorites.append(fingerprint)
            UserDefaults.standard.set(favorites, forKey: favoritesKey)
        }
    }

    // Helper method to remove a favorite (not exposed to JS, for internal use if needed)
    private func removeFavorite(_ fingerprint: String) {
        var favorites = UserDefaults.standard.array(forKey: favoritesKey) as? [String] ?? []
        if let index = favorites.firstIndex(of: fingerprint) {
            favorites.remove(at: index)
            UserDefaults.standard.set(favorites, forKey: favoritesKey)
        }
    }
}

File: android/src/main/java/com/reactnativebitchat/BitchatModule.java
package com.reactnativebitchat;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableNativeArray;

import android.content.SharedPreferences;
import java.util.Arrays;
import java.util.Set;

public class BitchatModule extends ReactContextBaseJavaModule {
    private final Channel channel = new Channel();
    private final BleService bleService = new BleService();
    private final Encryption encryption = new Encryption();
    private Callback messageCallback, peerConnectedCallback, peerDisconnectedCallback, deliveryAckCallback,
            readReceiptCallback, deliveryStatusUpdateCallback;
    private static final String PREFS_NAME = "BitchatPrefs";
    private static final String FAVORITES_KEY = "Favorites";

    public BitchatModule(ReactApplicationContext context) {
        super(context);
    }

    @Override
    public String getName() {
        return "Bitchat";
    }

    @ReactMethod
    public void startAdvertising(String peerID, Promise promise) {
        if (peerID.length() != 8) {
            promise.reject("INVALID_PEER_ID", "Peer ID must be 8 bytes");
            return;
        }
        try {
            bleService.startAdvertising(peerID);
            bleService.sendCoverTraffic(peerID); // Include cover traffic
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("ADVERTISE_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void scanPeers(Promise promise) {
        try {
            String[] peers = bleService.scanPeers();
            WritableArray result = new WritableNativeArray();
            for (String peer : peers) result.pushString(peer);
            promise.resolve(result);
        } catch (Exception e) {
            promise.reject("SCAN_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void sendMessage(ReadableMap message, String recipient, Promise promise) {
        try {
            BitchatMessage msg = mapToBitchatMessage(message);
            msg.deliveryStatus = "PENDING";
            if (recipient != null) msg.recipientNickname = recipient;
            bleService.sendMessage(msg, recipient);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("SEND_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void sendEncryptedChannelMessage(ReadableMap message, String channel, Promise promise) {
        try {
            BitchatMessage msg = mapToBitchatMessage(message);
            byte[] key = encryption.deriveChannelKey(channel, channel); // Use channel as password for simplicity
            msg.encryptedContent = encryption.encryptContent(msg.content, key);
            msg.isEncrypted = true;
            msg.channel = channel;
            bleService.sendEncryptedChannelMessage(msg, channel);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("ENCRYPT_SEND_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void onMessageReceived(Callback callback) {
        messageCallback = callback;
    }

    @ReactMethod
    public void onPeerConnected(Callback callback) {
        peerConnectedCallback = callback;
    }

    @ReactMethod
    public void onPeerDisconnected(Callback callback) {
        peerDisconnectedCallback = callback;
    }

    @ReactMethod
    public void onDeliveryAck(Callback callback) {
        deliveryAckCallback = callback;
    }

    @ReactMethod
    public void onReadReceipt(Callback callback) {
        readReceiptCallback = callback;
    }

    @ReactMethod
    public void onDeliveryStatusUpdate(Callback callback) {
        deliveryStatusUpdateCallback = callback;
    }

    @ReactMethod
    public void decryptChannelMessage(String encryptedContent, String channel, Promise promise) {
        try {
            byte[] key = encryption.deriveChannelKey(channel, channel); // Use channel as password
            String decrypted = encryption.decryptContent(encryptedContent.getBytes(), key);
            promise.resolve(decrypted);
        } catch (Exception e) {
            promise.reject("DECRYPT_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void isFavorite(String fingerprint, Promise promise) {
        if (fingerprint == null || fingerprint.isEmpty()) {
            promise.reject("INVALID_FINGERPRINT", "Fingerprint cannot be empty");
            return;
        }
        SharedPreferences prefs = getReactApplicationContext().getSharedPreferences(PREFS_NAME, 0);
        Set<String> favorites = prefs.getStringSet(FAVORITES_KEY, null);
        boolean isFavorite = favorites != null && favorites.contains(fingerprint);
        promise.resolve(isFavorite);
    }

    @ReactMethod
    public void joinChannel(String channel, String password, Promise promise) {
        try {
            boolean success = channel.joinChannel(channel, password, "peerID");
            promise.resolve(success);
        } catch (Exception e) {
            promise.reject("JOIN_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void setChannelPassword(String channel, String password, Promise promise) {
        try {
            channel.setChannelPassword(channel, password, "peerID");
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("SET_PASSWORD_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void removeChannelPassword(String channel, Promise promise) {
        try {
            channel.removeChannelPassword(channel, "peerID");
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("REMOVE_PASSWORD_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void transferChannelOwnership(String channel, String newOwnerID, Promise promise) {
        try {
            channel.transferOwnership(channel, newOwnerID, "peerID");
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("TRANSFER_ERROR", e.getMessage());
        }
    }

    private BitchatMessage mapToBitchatMessage(ReadableMap map) {
        BitchatMessage msg = new BitchatMessage();
        msg.id = map.getString("id");
        msg.sender = map.getString("sender");
        msg.content = map.getString("content");
        msg.timestamp = map.getDouble("timestamp");
        msg.isRelay = map.getBoolean("isRelay");
        if (map.hasKey("originalSender")) msg.originalSender = map.getString("originalSender");
        msg.isPrivate = map.getBoolean("isPrivate");
        if (map.hasKey("recipientNickname")) msg.recipientNickname = map.getString("recipientNickname");
        msg.senderPeerID = map.getString("senderPeerID");
        // Handle mentions, channel, encryptedContent, isEncrypted, deliveryStatus as needed
        return msg;
    }
}